unison-core/src/Unison/ABT.hs

```diff
@@ -244,12 +244,25 @@ into' a abt = case abt of
 
 -- | renames `old` to `new` in the given term, ignoring subtrees that bind `old`
 rename :: (Foldable f, Functor f, Var v) => v -> v -> Term f v a -> Term f v a
-rename old new t0@(Term _ ann t) = case t of
-  Var v -> if v == old then annotatedVar ann new else t0
-  Cycle body -> cycle' ann (rename old new body)
-  Abs v body -> if v == old then abs' ann v body
-                else abs' ann v (rename old new body)
-  Tm v -> tm' ann (fmap (rename old new) v)
+rename old new t0@(Term fvs ann t) =
+  if Set.notMember old fvs then t0
+  else case t of
+    Var v -> if v == old then annotatedVar ann new else t0
+    Cycle body -> cycle' ann (rename old new body)
+    Abs v body ->
+      -- v shadows old, so skip this subtree
+      if v == old then abs' ann v body
+
+      -- the rename would capture new, freshen this Abs
+      -- to make that no longer true, then proceed with
+      -- renaming `old` to `new`
+      else if v == new then
+        let v' = freshIn (Set.fromList [new,old] <> freeVars body) v
+        in abs' ann v' (rename old new (rename v v' body))
```

Comment:
@dolio This is the case I added to deal specifically with the case where `v == new`.

Idea for later, I think you mentioned this at some point - would be more efficient to have a rename operation that took a `Map v v`  or some such, rather than nesting the calls like this.

Reply:
I implemented simultaneous substitution in the normalized version of ABTs in my new-run branch, so we can port it back to normal ABTs after that gets merged or something.

Reply:
Great sounds perfect.